### ISSUE TITLE
Disable bound checking for functions in BoundaryConditions when using auto scheduling

### ISSUE THREAD

**Issue by lzhangzz** [halide/Halide#2846]
Hi, 

I'm writing a Generator that does bilinear interpolation when resizing an image.  The manual scheduled version worked fine. But when using the auto-scheduler I got the following error
```
Error:
The pure definition  of Function output calls function repeat_edge in an unbounded way in dimension 0
```
I have to clamp `oxf` and `oyf` to get auto scheduling working. It is weird that bound checking are applied to `BoundaryConditions` since they are actually used to set bounds when accessing a buffer. Is there a way to disable bound checking for `BoundaryConditons ` or it is designed like this intentionally?

```C++
class BilinearGenerator: public Halide::Generator<BilinearGenerator> {
 public:
  Input<Buffer<uint8_t>> input {"input", 3};
  Input<int> ow {"output_w"};
  Input<int> oh {"output_h"};
  Output<Buffer<uint8_t>> output {"output", 3};
  Var x{"x"}, y{"y"}, ox{"ox"}, oy{"oy"}, c{"c"};
  void generate() {
    Func clamped("clamped");
    clamped = BoundaryConditions::repeat_edge(input);
    auto rw = input.width();
    auto rh = input.height();
    Func oxf("oxf"), oyf("oyf"), ix("ix"), iy("iy"), ix0("ix0"), iy0("iy0"), ix1("ix1"), iy1("iy1"), wx("wx"), wy("wy");
    
    // normalized coords in output space
    oxf(ox, oy) = (ox + .5f) / ow;
    oyf(ox, oy) = (oy + .5f) / oh;
    // This will work
    // oxf(ox, oy) = clamp((ox + .5f) / ow, 0, 1);
    // oyf(ox, oy) = clamp((oy + .5f) / oh, 0, 1);
    
    // unnormalized coords in input space
    ix(ox, oy) = oxf(ox, oy) * rw - .5f;
    iy(ox, oy) = oyf(ox, oy) * rh - .5f;
    // coords of 4 pixels involved
    ix0(ox, oy) = cast<int>(floor(ix(ox, oy)));
    iy0(ox, oy) = cast<int>(floor(iy(ox, oy)));
    ix1(ox, oy) = cast<int>(ceil(ix(ox, oy)));
    iy1(ox, oy) = cast<int>(ceil(iy(ox, oy)));
    // interpolation weights
    wx(ox, oy) = ix(ox, oy) - ix0(ox, oy);
    wy(ox, oy) = iy(ox, oy) - iy0(ox, oy);
    Func p00("p00"), p01("p01"), p10("p10"), p11("p11");
    Func vx0("vx0"), vx1("vx1"), dst("dst");
    // load values
    p00(ox, oy, c) = clamped(ix0(ox, oy), iy0(ox, oy), c);
    p01(ox, oy, c) = clamped(ix1(ox, oy), iy0(ox, oy), c);
    p10(ox, oy, c) = clamped(ix0(ox, oy), iy1(ox, oy), c);
    p11(ox, oy, c) = clamped(ix1(ox, oy), iy1(ox, oy), c);
    // interpolate
    vx0(ox, oy, c) = lerp(p00(ox, oy, c), p01(ox, oy, c), wx(ox, oy));
    vx1(ox, oy, c) = lerp(p10(ox, oy, c), p11(ox, oy, c), wx(ox, oy));
    dst(ox, oy, c) = lerp(vx0(ox, oy, c), vx1(ox, oy, c), wy(ox, oy));
    output(ox, oy, c) = cast<uint8_t>(dst(ox, oy, c));
  }
  void schedule() {
    if (auto_schedule) {
      std::cout << "auto schedule." << std::endl;
      input.dim(0).set_bounds_estimate(0, 1024);
      input.dim(1).set_bounds_estimate(0, 1024);
      input.dim(2).set_bounds_estimate(0, 3);

      ow.set_estimate(256);
      oh.set_estimate(256);

      output.estimate(ox, 0, 1024).estimate(oy, 0, 1024).estimate(c, 0, 3);
    } else {
      std::cout << "manual schedule." << std::endl;
      output.reorder(c, ox, oy).unroll(c, 3).vectorize(ox, 32);
      output.print_loop_nest();
    }
  }
};
```

**Comment by zvookin** [comment]
If storage is scheduled for the result of the boundary condition, and the bounds with which is accessed cannot be determined, then this will happen. Bounds are also problematic as I don't think they are carried across Func boundaries fully, so this is perhaps even likely.

**Comment by psuriana** [comment]
Fixed in #3226
